### ISSUE TITLE
[JENKINS-26354] First time build does not show changelog

### ISSUE THREAD

**Issue by jenkins-infra-bot** [jenkinsci/git-plugin#2588]
<p>Suppose, we have <em>master</em> branch. Then we checkout to a new branch, lets name it <em>branch1</em>:</p>

<pre>
git checkout -b branch1
</pre>
<p>Then we do some commits:</p>

<pre>
echo <span class="code-quote">"a"</span> > a.txt && git add a.txt && git commit -m <span class="code-quote">'Added file: a.txt'</span>
echo <span class="code-quote">"b"</span> > b.txt && git add b.txt && git commit -m <span class="code-quote">'Added file: b.txt'</span>
</pre>
<p>Then we push our branch to Git repository:</p>

<pre>
git push origin branch1
</pre>
<p>In Jenkins, Git polling reports:</p>

<pre>
...
Seen branch in repository origin/branch1
Seen XXX remote branches
 > /usr/bin/git log --full-history --no-abbrev --format=raw -M -m --raw 0a6eb107fea5bb6371c450db1b5f6e100e0fba28..d8c5018ca184fb75e46386de2f16e552906fe106 # timeout=10
Done. Took 3.5 sec
Changes found
</pre>
<p>but Changes section shows nothing, it means, that no changes were made. Nevertheless, build was triggered.</p>

<p>The problem is deeper, suppose our build runs unit tests, and they failed. In that case, no emails will be send to committers, because of no changes. We have caught this case and got a problems <img class="emoticon" src="https://issues.jenkins.io/images/icons/emoticons/smile.png" height="16" width="16" align="absmiddle" alt="" border="0"/><br/>
I think, it's a bug, because 2 new commits were created and Jenkins should show it in Changes section.</p>

---
<details><summary><i>Originally reported by <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=maxusachev">maxusachev</a>, imported from: <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-26354" target="_blank">First time build does not show changelog</a></i></summary>
<i><ul>
<li><b>assignee</b>: <img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/ndeloof.png" title="ndeloof's avatar" /> <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=ndeloof">ndeloof</a>
<li><b>status</b>: Closed
<li><b>priority</b>: Major
<li><b>component(s)</b>: git-plugin
<li><b>label(s)</b>: git, git-plugin, polling, pollscm
<li><b>resolution</b>: Won't Fix
<li><b>resolved</b>: 2015-01-17T17:23:56+00:00
<li><b>votes</b>: 0
<li><b>watchers</b>: 17
<li><b>imported</b>: 2025-12-02
</ul></i>
<details><summary>Raw content of original issue</summary>

<pre>
<p>Suppose, we have <em>master</em> branch. Then we checkout to a new branch, lets name it <em>branch1</em>:</p>
<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java">git checkout -b branch1
</pre>
</div></div>
<p>Then we do some commits:</p>
<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java">echo <span class="code-quote">"a"</span> &gt; a.txt &amp;&amp; git add a.txt &amp;&amp; git commit -m <span class="code-quote">'Added file: a.txt'</span>
echo <span class="code-quote">"b"</span> &gt; b.txt &amp;&amp; git add b.txt &amp;&amp; git commit -m <span class="code-quote">'Added file: b.txt'</span>
</pre>
</div></div>
<p>Then we push our branch to Git repository:</p>
<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java">git push origin branch1
</pre>
</div></div>
<p>In Jenkins, Git polling reports:</p>
<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java">...
Seen branch in repository origin/branch1
Seen XXX remote branches
 &gt; /usr/bin/git log --full-history --no-abbrev --format=raw -M -m --raw 0a6eb107fea5bb6371c450db1b5f6e100e0fba28..d8c5018ca184fb75e46386de2f16e552906fe106 # timeout=10
Done. Took 3.5 sec
Changes found
</pre>
</div></div>
<p>but Changes section shows nothing, it means, that no changes were made. Nevertheless, build was triggered.</p>

<p>The problem is deeper, suppose our build runs unit tests, and they failed. In that case, no emails will be send to committers, because of no changes. We have caught this case and got a problems <img class="emoticon" src="https://issues.jenkins.io/images/icons/emoticons/smile.png" height="16" width="16" align="absmiddle" alt="" border="0"/>
I think, it's a bug, because 2 new commits were created and Jenkins should show it in Changes section.</p></pre>
</details>
</details>
<details><summary><i>environment</i></summary>

```
Jenkins ver. 1.591<br/>
Git plugin 2.3.3<br/>
GIT client plugin 1.15.0
```
</details>
<details><summary><i>1 attachment</i></summary>

- [addbeh.png](https://raw.githubusercontent.com/jenkinsci/attachments-from-jira-issues-misc/refs/heads/main/attachments/40436/addbeh.png)
  > ![addbeh.png](https://raw.githubusercontent.com/jenkinsci/attachments-from-jira-issues-misc/refs/heads/main/attachments/40436/addbeh.png)
</details>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/markewaite.png" title="markewaite's avatar" /> markewaite:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-26354?focusedId=219045&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-219045">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>I can't duplicate the bug based on the steps you've provided. Can you provide more details or clarify the areas in which my steps are different than your steps?</p>

<p>Steps I took:</p>

<ol>
	<li>Create a new git repo ( repo=~/git/bugs/<a href="https://issues.jenkins.io/browse/JENKINS-26534" title="Cannot run tests and set custom arguments" class="issue-link" data-issue-key="JENKINS-26534">JENKINS-26534</a>; mkdir -p $repo; cd $repo; git init )</li>
	<li>Commit a change to that repo ( branch=master; date &gt;&gt; $branch; git add $branch; git commit -m "Add date to $branch" )</li>
	<li>Create a freestyle project to monitor that repo ( poll git often, or create a post-receive hook )</li>
	<li>Define "Branches to build" as "*" so that any branch which appears will start a build</li>
	<li>Confirm that the project builds on its first polling interval</li>
	<li>Confirm that the log of the first build contains the line "First time build. Skipping changelog."</li>
	<li>Confirm that Changes and Recent Changes show no change for that first time build</li>
	<li>Add a branch to the repo with a commit ( branch=branch-a; git checkout -b $branch; date &gt;&gt; $branch; git add $branch; git commit -m "Add date to $branch" )</li>
	<li>Confirm that the next poll starts that job</li>
	<li>Confirm that the console output contains the line "First time build. Skipping changelog."</li>
	<li>Commit a second change to the branch (branch=branch-a; git checkout $branch; date &gt;&gt; $branch; git add $branch; git commit -m "Add date to $branch" )</li>
	<li>Confirm that the next poll starts that job</li>
	<li>Confirm that the changes show the commit (since it is not a first time build)</li>
</ol>


<p>Is the bug you're reporting that the first time build for a new branch does not list any changes? If so, it is difficult to make a good general case decision of what should be the basis of those changes.</p>
</pre>
</details>

</ul>
</details>

<p>I can't duplicate the bug based on the steps you've provided. Can you provide more details or clarify the areas in which my steps are different than your steps?</p>

<p>Steps I took:</p>

<ol>
	<li>Create a new git repo ( repo=~/git/bugs/<a href="https://issues.jenkins.io/browse/JENKINS-26534" title="Cannot run tests and set custom arguments" class="issue-link" data-issue-key="JENKINS-26534">JENKINS-26534</a>; mkdir -p $repo; cd $repo; git init )</li>
	<li>Commit a change to that repo ( branch=master; date >> $branch; git add $branch; git commit -m "Add date to $branch" )</li>
	<li>Create a freestyle project to monitor that repo ( poll git often, or create a post-receive hook )</li>
	<li>Define "Branches to build" as "*" so that any branch which appears will start a build</li>
	<li>Confirm that the project builds on its first polling interval</li>
	<li>Confirm that the log of the first build contains the line "First time build. Skipping changelog."</li>
	<li>Confirm that Changes and Recent Changes show no change for that first time build</li>
	<li>Add a branch to the repo with a commit ( branch=branch-a; git checkout -b $branch; date >> $branch; git add $branch; git commit -m "Add date to $branch" )</li>
	<li>Confirm that the next poll starts that job</li>
	<li>Confirm that the console output contains the line "First time build. Skipping changelog."</li>
	<li>Commit a second change to the branch (branch=branch-a; git checkout $branch; date >> $branch; git add $branch; git commit -m "Add date to $branch" )</li>
	<li>Confirm that the next poll starts that job</li>
	<li>Confirm that the changes show the commit (since it is not a first time build)</li>
</ol>


<p>Is the bug you're reporting that the first time build for a new branch does not list any changes? If so, it is difficult to make a good general case decision of what should be the basis of those changes.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/markewaite.png" title="markewaite's avatar" /> markewaite:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-26354?focusedId=219054&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-219054">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>I confirmed that a multi-branch job (recently added plugin which allows a git or subversion based job to create a new job for each new branch detected) behaves similarly. The first build of a new job lists in the console output that it is a first time build and is not computing the changelog. Subsequent builds compute the changelog.</p>
</pre>
</details>

</ul>
</details>

<p>I confirmed that a multi-branch job (recently added plugin which allows a git or subversion based job to create a new job for each new branch detected) behaves similarly. The first build of a new job lists in the console output that it is a first time build and is not computing the changelog. Subsequent builds compute the changelog.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>maxusachev:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-26354?focusedId=219065&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-219065">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Yes, the problem related to the first time build for new branch, which was discovered by Git polling. For the next changes in existing remote branch changes are shown correct.
I don't know details of the implementation of Git plugin, but it seems to me, that plugin can store list of all commit hashes of repository and when it see new remote branch, try to find those commits which are in new branch but not in commits store.
And then build changelog based on these commits.</p>
</pre>
</details>

</ul>
</details>

<p>Yes, the problem related to the first time build for new branch, which was discovered by Git polling. For the next changes in existing remote branch changes are shown correct.<br/>
I don't know details of the implementation of Git plugin, but it seems to me, that plugin can store list of all commit hashes of repository and when it see new remote branch, try to find those commits which are in new branch but not in commits store.<br/>
And then build changelog based on these commits.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/markewaite.png" title="markewaite's avatar" /> markewaite:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-26354?focusedId=219581&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-219581">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>There isn't any way to reliably decide which branch should be the basis for the diff calculation you're proposing for an initial branch. There may be many different heuristics (fallible rules) which could be applied to the case, but there will be cases where each of the heuristics is wrong.</p>
</pre>
</details>

</ul>
</details>

<p>There isn't any way to reliably decide which branch should be the basis for the diff calculation you're proposing for an initial branch. There may be many different heuristics (fallible rules) which could be applied to the case, but there will be cases where each of the heuristics is wrong.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>benja1:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-26354?focusedId=295160&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-295160">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>This happens for us too. The git plugin reports the correct commit hash of the changeset but it shows No Changes on the Changes view.</p>

<p>What is the fix? Thanks.</p>
</pre>
</details>

</ul>
</details>

<p>This happens for us too. The git plugin reports the correct commit hash of the changeset but it shows No Changes on the Changes view.</p>

<p>What is the fix? Thanks.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/markewaite.png" title="markewaite's avatar" /> markewaite:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-26354?focusedId=295163&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-295163">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=benja1" class="user-hover" rel="benja1">benja1</a>, as far as I can tell from this bug report, it wants that the first commit on a newly created branch should show changes from something, but "something" can't be reliably determined.  Are you requesting that the first build on a new branch should show changes from the branch that was used as the basis for the new branch?</p>

<p>There is an "Additional Behaviour" which will allow all change reports to be based on a specific branch.  If your workflow is creating branches from a specific, single base branch, you could list that base branch as the base to compute changes.</p>

<p>There isn't a general way to decide which branch is the "origin" for the current commit, since it is a directed graph from the current commit to its predecessors.  There could be many different branches in that directed graph, and the "real" base branch could have advanced since the branch point.  I'm quite hesitant to place a "parent guesser" into the plugin, since that will result in some set of users being frustrated that it did not guess their case.</p>

<p>The bug was resolved as "Won't fix"...</p>
</pre>
</details>

</ul>
</details>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=benja1" class="user-hover" rel="benja1">benja1</a>, as far as I can tell from this bug report, it wants that the first commit on a newly created branch should show changes from something, but "something" can't be reliably determined.  Are you requesting that the first build on a new branch should show changes from the branch that was used as the basis for the new branch?</p>

<p>There is an "Additional Behaviour" which will allow all change reports to be based on a specific branch.  If your workflow is creating branches from a specific, single base branch, you could list that base branch as the base to compute changes.</p>

<p>There isn't a general way to decide which branch is the "origin" for the current commit, since it is a directed graph from the current commit to its predecessors.  There could be many different branches in that directed graph, and the "real" base branch could have advanced since the branch point.  I'm quite hesitant to place a "parent guesser" into the plugin, since that will result in some set of users being frustrated that it did not guess their case.</p>

<p>The bug was resolved as "Won't fix"...</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>benja1:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-26354?focusedId=295165&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-295165">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Exactly. We perform a first commit(s) and push of a new branch. The Jenkins job is set to build any branch in the repository with the "**" branch specifier.</p>

<p>To verify, is this a problem with the git plugin? Is the issue tracked anywhere else? This feature is important to our workflow.</p>
</pre>
</details>

</ul>
</details>

<p>Exactly. We perform a first commit(s) and push of a new branch. The Jenkins job is set to build any branch in the repository with the "**" branch specifier.</p>

<p>To verify, is this a problem with the git plugin? Is the issue tracked anywhere else? This feature is important to our workflow.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/markewaite.png" title="markewaite's avatar" /> markewaite:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-26354?focusedId=295166&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-295166">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>It is tracked here, and has been decided that it won't be fixed.</p>

<p>It is how the git plugin is defined to behave.  Unless the "base branch to compute changes" is set, the first build does not attempt to compute a difference with any predecessor, because there is no predecessor for the first build on a branch.</p>
</pre>
</details>

</ul>
</details>

<p>It is tracked here, and has been decided that it won't be fixed.</p>

<p>It is how the git plugin is defined to behave.  Unless the "base branch to compute changes" is set, the first build does not attempt to compute a difference with any predecessor, because there is no predecessor for the first build on a branch.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>benja1:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-26354?focusedId=295167&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-295167">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Is there a way to work around this issue? We also do not want our Jenkins job to build on branch creation when the branch is first pushed to the remote repository. We are only creating local branches and then pushing them once the feature is complete because otherwise the Jenkins job is triggered upon branch creation with no commits yet.</p>
</pre>
</details>

</ul>
</details>

<p>Is there a way to work around this issue? We also do not want our Jenkins job to build on branch creation when the branch is first pushed to the remote repository. We are only creating local branches and then pushing them once the feature is complete because otherwise the Jenkins job is triggered upon branch creation with no commits yet.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/markewaite.png" title="markewaite's avatar" /> markewaite:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-26354?focusedId=295168&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-295168">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>You might push a "dummy commit" as the first commit on a new branch so that Jenkins will detect the new branch and see that it is not the same SHA1 as has already been built, then future changes after that "dummy commit" will show as differences in the Jenkins job.</p>
</pre>
</details>

</ul>
</details>

<p>You might push a "dummy commit" as the first commit on a new branch so that Jenkins will detect the new branch and see that it is not the same SHA1 as has already been built, then future changes after that "dummy commit" will show as differences in the Jenkins job.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>pajok:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-26354?focusedId=320229&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-320229">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=markewaite" class="user-hover" rel="markewaite">markewaite</a> you told that there is some "Additional Behaviour which will allow all change reports to be based on a specific branch". Could you please tell me which exactly Additional Behaviour could do that.</p>

<p>I am using Git Plugin v.3.6.4, and all behaviour that i have are shown on picture.</p>

<p>Thx. <span class="image-wrap" style=""><img src="https://issues.jenkins.io/secure/attachment/40436/40436_addbeh.png" style="border: 0px solid black" /></span></p>
</pre>
</details>

</ul>
</details>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=markewaite" class="user-hover" rel="markewaite">markewaite</a> you told that there is some "Additional Behaviour which will allow all change reports to be based on a specific branch". Could you please tell me which exactly Additional Behaviour could do that.</p>

<p>I am using Git Plugin v.3.6.4, and all behaviour that i have are shown on picture.</p>

<p>Thx. <span class="image-wrap" style=""><img src="https://raw.githubusercontent.com/jenkinsci/attachments-from-jira-issues-misc/refs/heads/main/attachments/40436/addbeh.png" style="border: 0px solid black" /></span></p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/markewaite.png" title="markewaite's avatar" /> markewaite:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-26354?focusedId=320252&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-320252">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=pajok" class="user-hover" rel="pajok">pajok</a> the option is not available for multi-branch pipelines.  The first build of a multi-branch pipeline happens when the branch is discovered.  Multi-branch pipeline checkout has intentionally simplified the available options to those needed by most users.</p>

<p>I suspect you could create a plugin which adds that option.  That has the benefit that typical users don't see the option unless they choose to install the plugin.</p>
</pre>
</details>

</ul>
</details>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=pajok" class="user-hover" rel="pajok">pajok</a> the option is not available for multi-branch pipelines.  The first build of a multi-branch pipeline happens when the branch is discovered.  Multi-branch pipeline checkout has intentionally simplified the available options to those needed by most users.</p>

<p>I suspect you could create a plugin which adds that option.  That has the benefit that typical users don't see the option unless they choose to install the plugin.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>olegstepura:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-26354?focusedId=364442&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-364442">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Hi!</p>

<p>Sorry to resurrect this old issue, but we faced same issue and need some way to solve it. We use multibranch pipeline with scripted Jenkinsfile which should run some stages based on changes we have in a github PR. For this reason we use <tt>currentBuild.changeSets</tt> and continue collecting changes up to first commit in a branch (until <tt>build.previousBuild</tt> is null, in a while loop). But this strange issue was discovered and I wonder what are the options to solve this nowadays?</p>

<p>Seems like we still don't have a "hidden" option to define parent branch, but I wonder why it is a problem actually. Github solves this somehow in their "commits" Tab. Doesn't any GIT commit point to it's parent commit? Cannot we compare using this info?</p>

<p> </p>

<p>Thanks!</p>
</pre>
</details>

</ul>
</details>

<p>Hi!</p>

<p>Sorry to resurrect this old issue, but we faced same issue and need some way to solve it. We use multibranch pipeline with scripted Jenkinsfile which should run some stages based on changes we have in a github PR. For this reason we use <tt>currentBuild.changeSets</tt> and continue collecting changes up to first commit in a branch (until <tt>build.previousBuild</tt> is null, in a while loop). But this strange issue was discovered and I wonder what are the options to solve this nowadays?</p>

<p>Seems like we still don't have a "hidden" option to define parent branch, but I wonder why it is a problem actually. Github solves this somehow in their "commits" Tab. Doesn't any GIT commit point to it's parent commit? Cannot we compare using this info?</p>

<p> </p>

<p>Thanks!</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/markewaite.png" title="markewaite's avatar" /> markewaite:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-26354?focusedId=364445&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-364445">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=olegstepura" class="user-hover" rel="olegstepura">olegstepura</a> <a href="https://github.com/jenkinsci/git-plugin/pull/637" class="external-link" target="_blank" rel="nofollow noopener">Pull request 637</a> includes a proposal to provide a changelog for the first build of a branch.  It is currently in the "Later" milestone because it is not receiving my focused evaluation right now.  After git plugin 4.0 has stabilized and released, I'll evaluate the  pull requests in the "Later" milestone to decide which should be reopened and be reviewed.</p>

<p>I'd love to have your help evaluating the changes that are proposed in that pull request.  You'd need to build a local copy of the plugin based on that pull request, use it in your environment, and report the results to the pull request.</p>

<p>If you're not comfortable building the pull request, I could temporarily reopen the request, have the CI servers build it, then point you to the built artifact so that you could test it.</p>

<p>See <a href="https://issues.jenkins.io/browse/JENKINS-14138" title="New git branch has empty changelist" class="issue-link" data-issue-key="JENKINS-14138">JENKINS-14138</a> and that pull request for more details</p>
</pre>
</details>

</ul>
</details>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=olegstepura" class="user-hover" rel="olegstepura">olegstepura</a> <a href="https://github.com/jenkinsci/git-plugin/pull/637" class="external-link" target="_blank" rel="nofollow noopener">Pull request 637</a> includes a proposal to provide a changelog for the first build of a branch.  It is currently in the "Later" milestone because it is not receiving my focused evaluation right now.  After git plugin 4.0 has stabilized and released, I'll evaluate the  pull requests in the "Later" milestone to decide which should be reopened and be reviewed.</p>

<p>I'd love to have your help evaluating the changes that are proposed in that pull request.  You'd need to build a local copy of the plugin based on that pull request, use it in your environment, and report the results to the pull request.</p>

<p>If you're not comfortable building the pull request, I could temporarily reopen the request, have the CI servers build it, then point you to the built artifact so that you could test it.</p>

<p>See <a href="https://issues.jenkins.io/browse/JENKINS-14138" title="New git branch has empty changelist" class="issue-link" data-issue-key="JENKINS-14138">JENKINS-14138</a> and that pull request for more details</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>olegstepura:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-26354?focusedId=364524&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-364524">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Hi!</p>

<p>Thanks for quick reply and sorry I did not react as fast. I discussed internally if this is acceptable to do what you suggested and it was not, but another option to workaround the limitations here was found: <a href="https://issues.jenkins.io/browse/JENKINS-54285" title="Changeset not reflective of Pull Request" class="issue-link" data-issue-key="JENKINS-54285">JENKINS-54285</a></p>

<p>I tried using commands from that ticket, it works if we use big enough shallow clone depth (I set it to 100).</p>

<p>the only issue I faced was being unable to use <tt>sh</tt> in a function. It looks buggy. Seems like I cannot simply define a function like this:</p>

<p> </p>
<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java">def getLastCommitHash() {
  <span class="code-keyword">return</span> sh(
    returnStdout: <span class="code-keyword">true</span>,
    script: <span class="code-quote">"git show -s --no-abbrev-commit --pretty=format:%P%n%H%n HEAD"</span>
  ).trim().split(<span class="code-quote">"\n"</span>).last().trim()
}</pre>
</div></div>
<p> </p>

<p>and the use it: </p>
<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java">def otherFunction() {
  def lastCommit = getLastCommitHash()
  <span class="code-comment">// <span class="code-keyword">do</span> something <span class="code-keyword">else</span> with it expecting lastCommit has one-line string
</span>  <span class="code-keyword">return</span> [a, b, c]
}</pre>
</div></div>
<p>In the example above <tt>otherFunction()</tt> will return exact unprocessed output of <tt>sh</tt> command from <tt>getLastCommitHash()</tt> function. This is weird.</p>

<p>I was able to workaround this creating simple functions which only run <tt>sh</tt> commands and do no post-processing (just return <tt>sh</tt> output) and other functions that only work with strings and not do anything with <tt>sh</tt>. Then all those functions are used together inside our main <tt>node {</tt>} context.</p>
<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java">def getHeads() {
  <span class="code-keyword">return</span> sh(
    returnStdout: <span class="code-keyword">true</span>,
    script: <span class="code-quote">"git show -s --no-abbrev-commit --pretty=format:%P%n%H%n HEAD"</span>
  )
}

def getLastCommitHash(headsOutput) {
  <span class="code-keyword">return</span> headsOutput.trim().split(<span class="code-quote">"\n"</span>).last().trim()
}</pre>
</div></div>
<p> then use this in <tt>Jenkinsfile:</tt></p>
<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java">node(<span class="code-quote">"xxx"</span>) {
  def headsOutput = getHeads()
  def lastCommitHash = getLastCommitHash(headsOutput)
  <span class="code-comment">// ...
</span>}</pre>
</div></div>
<p> </p>
</pre>
</details>

</ul>
</details>

<p>Hi!</p>

<p>Thanks for quick reply and sorry I did not react as fast. I discussed internally if this is acceptable to do what you suggested and it was not, but another option to workaround the limitations here was found: <a href="https://issues.jenkins.io/browse/JENKINS-54285" title="Changeset not reflective of Pull Request" class="issue-link" data-issue-key="JENKINS-54285">JENKINS-54285</a></p>

<p>I tried using commands from that ticket, it works if we use big enough shallow clone depth (I set it to 100).</p>

<p>the only issue I faced was being unable to use <tt>sh</tt> in a function. It looks buggy. Seems like I cannot simply define a function like this:</p>

<p> </p>

<pre>
def getLastCommitHash() {
  <span class="code-keyword">return</span> sh(
    returnStdout: <span class="code-keyword">true</span>,
    script: <span class="code-quote">"git show -s --no-abbrev-commit --pretty=format:%P%n%H%n HEAD"</span>
  ).trim().split(<span class="code-quote">"\n"</span>).last().trim()
}</pre>
<p> </p>

<p>and the use it: </p>

<pre>
def otherFunction() {
  def lastCommit = getLastCommitHash()
  <span class="code-comment">// <span class="code-keyword">do</span> something <span class="code-keyword">else</span> with it expecting lastCommit has one-line string
</span>  <span class="code-keyword">return</span> [a, b, c]
}</pre>
<p>In the example above <tt>otherFunction()</tt> will return exact unprocessed output of <tt>sh</tt> command from <tt>getLastCommitHash()</tt> function. This is weird.</p>

<p>I was able to workaround this creating simple functions which only run <tt>sh</tt> commands and do no post-processing (just return <tt>sh</tt> output) and other functions that only work with strings and not do anything with <tt>sh</tt>. Then all those functions are used together inside our main <tt>node {</tt>} context.</p>

<pre>
def getHeads() {
  <span class="code-keyword">return</span> sh(
    returnStdout: <span class="code-keyword">true</span>,
    script: <span class="code-quote">"git show -s --no-abbrev-commit --pretty=format:%P%n%H%n HEAD"</span>
  )
}

def getLastCommitHash(headsOutput) {
  <span class="code-keyword">return</span> headsOutput.trim().split(<span class="code-quote">"\n"</span>).last().trim()
}</pre>
<p> then use this in <tt>Jenkinsfile:</tt></p>

<pre>
node(<span class="code-quote">"xxx"</span>) {
  def headsOutput = getHeads()
  def lastCommitHash = getLastCommitHash(headsOutput)
  <span class="code-comment">// ...
</span>}</pre>
<p> </p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>tamerlaha:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-26354?focusedId=373116&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-373116">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Is that really resolved?</p>

<p>The problem is really serious  we can't use 'changeset' in when condition with branches at all.</p>
</pre>
</details>

</ul>
</details>

<p>Is that really resolved?</p>

<p>The problem is really serious  we can't use 'changeset' in when condition with branches at all.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/markewaite.png" title="markewaite's avatar" /> markewaite:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-26354?focusedId=373127&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-373127">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>It is resolved as "Won't Fix".  Refer to <a href="https://issues.jenkins-ci.org/browse/JENKINS-26354?focusedCommentId=364445&amp;page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-364445" class="external-link" target="_blank" rel="nofollow noopener">the earlier discussions</a> for more details.</p>

<p>There is <a href="https://github.com/jenkinsci/git-plugin/pull/637" class="external-link" target="_blank" rel="nofollow noopener">this pull request</a> that attempts to apply heuristics to decide which change should be used as the basis for a changeset.  You're welcome to propose further changes to it to resolve the issues that are discussed in it.</p>
</pre>
</details>

</ul>
</details>

<p>It is resolved as "Won't Fix".  Refer to <a href="https://issues.jenkins-ci.org/browse/JENKINS-26354?focusedCommentId=364445&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-364445" class="external-link" target="_blank" rel="nofollow noopener">the earlier discussions</a> for more details.</p>

<p>There is <a href="https://github.com/jenkinsci/git-plugin/pull/637" class="external-link" target="_blank" rel="nofollow noopener">this pull request</a> that attempts to apply heuristics to decide which change should be used as the basis for a changeset.  You're welcome to propose further changes to it to resolve the issues that are discussed in it.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>azelezni:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-26354?focusedId=373208&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-373208">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>I was able to get the changeset using the following:</p>


<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java">checkout scm: [
    $class: <span class="code-quote">'GitSCM'</span>,
    branches: [[name: env.BRANCH_NAME]],
    extensions: [
        [
            $class: <span class="code-quote">'ChangelogToBranch'</span>,
            options: [
                compareRemote: <span class="code-quote">'origin'</span>,
                compareTarget: <span class="code-quote">'master'</span>
            ]
        ]
    ],
    userRemoteConfigs: [
        [
            url: scm.getUserRemoteConfigs()[0].getUrl(),
            refspec: scm.getUserRemoteConfigs()[0].getRefspec(),
            credentialsId: <span class="code-quote">'github-privatekey'</span>
        ]
    ]
]
</pre>
</div></div>
<p>With the above I can get a changeset from the first build.</p>

<p> </p>
</pre>
</details>

</ul>
</details>

<p>I was able to get the changeset using the following:</p>



<pre>
checkout scm: [
    $class: <span class="code-quote">'GitSCM'</span>,
    branches: [[name: env.BRANCH_NAME]],
    extensions: [
[
    $class: <span class="code-quote">'ChangelogToBranch'</span>,
    options: [
compareRemote: <span class="code-quote">'origin'</span>,
compareTarget: <span class="code-quote">'master'</span>
    ]
]
    ],
    userRemoteConfigs: [
[
    url: scm.getUserRemoteConfigs()[0].getUrl(),
    refspec: scm.getUserRemoteConfigs()[0].getRefspec(),
    credentialsId: <span class="code-quote">'github-privatekey'</span>
]
    ]
]
</pre>
<p>With the above I can get a changeset from the first build.</p>

<p> </p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/bjhaid.jpg" title="bjhaid's avatar" /> bjhaid:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-26354?focusedId=384367&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-384367">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=markewaite" class="user-hover" rel="markewaite">markewaite</a> what are your thoughts on having `changeset` support a remote ref variable with a default being an empty string, that way a user can specify the base remote ref and that can be used to help the plugin decide how to compute the changeset/changelog. The default variable ensures the code is backwards compatible so folks who don't explicitly supply a remote ref gets their old behavior. I have not dug into the code so I don't know if my suggestion is at all realistic.</p>
</pre>
</details>

</ul>
</details>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=markewaite" class="user-hover" rel="markewaite">markewaite</a> what are your thoughts on having `changeset` support a remote ref variable with a default being an empty string, that way a user can specify the base remote ref and that can be used to help the plugin decide how to compute the changeset/changelog. The default variable ensures the code is backwards compatible so folks who don't explicitly supply a remote ref gets their old behavior. I have not dug into the code so I don't know if my suggestion is at all realistic.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/markewaite.png" title="markewaite's avatar" /> markewaite:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-26354?focusedId=384386&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-384386">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=bjhaid" class="user-hover" rel="bjhaid">bjhaid</a> the plugin has the concept of a base branch for comparison.  Users that are willing to use the base branch to compute the changelog might be able to see changelogs on the first commit, though that changelog will be incorrect if their actual base branch does not match the base branch used to compute the changelog.</p>

<p>As an example, there are two development branches for a GitHub repo, one named 'master' and one named 'stable-3.x'.  If the Jenkinsfile declares the changelog should be computed against the 'master' branch, then all pull requests to the 'stable-3.x' branch will incorrectly compute their changelog against the 'master' branch.</p>
</pre>
</details>

</ul>
</details>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=bjhaid" class="user-hover" rel="bjhaid">bjhaid</a> the plugin has the concept of a base branch for comparison.  Users that are willing to use the base branch to compute the changelog might be able to see changelogs on the first commit, though that changelog will be incorrect if their actual base branch does not match the base branch used to compute the changelog.</p>

<p>As an example, there are two development branches for a GitHub repo, one named 'master' and one named 'stable-3.x'.  If the Jenkinsfile declares the changelog should be computed against the 'master' branch, then all pull requests to the 'stable-3.x' branch will incorrectly compute their changelog against the 'master' branch.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/bjhaid.jpg" title="bjhaid's avatar" /> bjhaid:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-26354?focusedId=384389&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-384389">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Thanks for the response!</p>

<p>&gt; If the Jenkinsfile declares the changelog should be computed against the 'master' branch, then all pull requests to the 'stable-3.x' branch will incorrectly compute their changelog against the 'master' branch.</p>

<p>I understand this concern, do you think documentation should be adequate in covering this concern? Also what are your thoughts on refinement of my suggestion below:</p>

<p>A user can provide a fall through base branch when the plugin cannot correctly compute the changelog (i.e on the first commit on a branch), obviously shares similar risks as the original proposal, but in this case it risk is much less, and will still need documentation on the risks.</p>
</pre>
</details>

</ul>
</details>

<p>Thanks for the response!</p>

<p>> If the Jenkinsfile declares the changelog should be computed against the 'master' branch, then all pull requests to the 'stable-3.x' branch will incorrectly compute their changelog against the 'master' branch.</p>

<p>I understand this concern, do you think documentation should be adequate in covering this concern? Also what are your thoughts on refinement of my suggestion below:</p>

<p>A user can provide a fall through base branch when the plugin cannot correctly compute the changelog (i.e on the first commit on a branch), obviously shares similar risks as the original proposal, but in this case it risk is much less, and will still need documentation on the risks.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>djviking:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-26354?focusedId=390114&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-390114">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>This issue has also been experienced on branches that had been rebased.
If you rebase N-commits, for which thos commits was a part of the last builds changeset, then it will treat the new build as a first time build.</p>

<p>Using ChangelogToBranch could work, but it will add additional boilerplate code to our pipeline, instead of just calling checkout scm.
For use it is useless to configure it for the job, because we do not know the base branch (compareTarget) until the job has started.</p>

<p>Does the git plugin check for changes using the git commit hash or message?</p>

<p>If it detects "first time build", then it should use the latest commit as changeset.
Build #1
&gt; Commit message: "fix: Correct this method"
&gt; First time build. Skipping changelog</p>

<p>Rebase commit, new hash, same message.
Build #2
&gt; Commit message: "fix: Correct this method"
&gt; First time build. Skipping changelog</p>
</pre>
</details>

</ul>
</details>

<p>This issue has also been experienced on branches that had been rebased.<br/>
If you rebase N-commits, for which thos commits was a part of the last builds changeset, then it will treat the new build as a first time build.</p>

<p>Using ChangelogToBranch could work, but it will add additional boilerplate code to our pipeline, instead of just calling checkout scm.<br/>
For use it is useless to configure it for the job, because we do not know the base branch (compareTarget) until the job has started.</p>

<p>Does the git plugin check for changes using the git commit hash or message?</p>

<p>If it detects "first time build", then it should use the latest commit as changeset.<br/>
Build #1<br/>
> Commit message: "fix: Correct this method"<br/>
> First time build. Skipping changelog</p>

<p>Rebase commit, new hash, same message.<br/>
Build #2<br/>
> Commit message: "fix: Correct this method"<br/>
> First time build. Skipping changelog</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/markewaite.png" title="markewaite's avatar" /> markewaite:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-26354?focusedId=390128&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-390128">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<blockquote><p>Does the git plugin check for changes using the git commit hash or message?</p></blockquote>

<p>It uses the commit hash.</p>
</pre>
</details>

</ul>
</details>

<blockquote><p>Does the git plugin check for changes using the git commit hash or message?</p></blockquote>

<p>It uses the commit hash.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>djviking:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-26354?focusedId=390597&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-390597">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Is it a bug then, when a rebase there are no changes? It is a changed commit and new hash, but it treats it as First time build thus no changes.</p>
</pre>
</details>

</ul>
</details>

<p>Is it a bug then, when a rebase there are no changes? It is a changed commit and new hash, but it treats it as First time build thus no changes.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/markewaite.png" title="markewaite's avatar" /> markewaite:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-26354?focusedId=390598&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-390598">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=djviking" class="user-hover" rel="djviking">djviking</a> I like the definition of "bug" that Gerald Weinberg and Cem Kaner offered.  It's a "bug" if it "bugs someone".  In the case you're describing, a bug report would be a good way to start the discussions of what you would expect as a user in the case of a rebase.</p>

<p>I believe the current process compares the preceding build to the current build.  If there is no path from the preceding build to current build (as would often happen in a rebase), then no changes are displayed.</p>

<p>If you decide to open that issue, please describe what you would like to happen as a user, paying particular attention to the cases where you're assuming that the git plugin knows the base branch from which the pull request started.  It doesn't actually know a base branch unless specifically told to show differences against a base branch.  Heuristics to guess the base branch are prone to fail in all sorts of unfriendly ways.</p>
</pre>
</details>

</ul>
</details>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=djviking" class="user-hover" rel="djviking">djviking</a> I like the definition of "bug" that Gerald Weinberg and Cem Kaner offered.  It's a "bug" if it "bugs someone".  In the case you're describing, a bug report would be a good way to start the discussions of what you would expect as a user in the case of a rebase.</p>

<p>I believe the current process compares the preceding build to the current build.  If there is no path from the preceding build to current build (as would often happen in a rebase), then no changes are displayed.</p>

<p>If you decide to open that issue, please describe what you would like to happen as a user, paying particular attention to the cases where you're assuming that the git plugin knows the base branch from which the pull request started.  It doesn't actually know a base branch unless specifically told to show differences against a base branch.  Heuristics to guess the base branch are prone to fail in all sorts of unfriendly ways.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>mkroell:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-26354?focusedId=399603&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-399603">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Hello, I am experiencing these issues as well using Bitbucket with a Multibranch Pipeline. This is causing issues with using other plugins looking to use the changelog, specifically the Atlassian Jira Cloud plugin.  Here is some logs on the issue:</p>

<p><span class="error">&#91;Pipeline&#93;</span> {
<span class="error">&#91;Pipeline&#93;</span> stage
<span class="error">&#91;Pipeline&#93;</span> </p>
{ (Declarative: Checkout SCM)
[Pipeline] checkout
The recommended git tool is: NONE
using credential Jenkins-Bitbucket-Build-Status-Notifier
Fetching changes from the remote Git repository
Fetching without tags
 &gt; git rev-parse --is-inside-work-tree # timeout=10
 &gt; git config remote.origin.url https://x-token-auth##@bitbucket.org/##/jenkins-experimental.git # timeout=10
Fetching upstream changes from https://x-token-auth@bitbucket.org/##/jenkins-experimental.git
 &gt; git --version # timeout=10
 &gt; git --version # 'git version 2.20.1'
 &gt; git fetch --no-tags --force --progress -- https://x-token-auth:##@bitbucket.org/##/jenkins-experimental.git +refs/heads/CDE-888-Jira-Test:refs/remotes/origin/CDE-888-Jira-Test # timeout=10
Checking out Revision a40df9adb4445594d996cb282389acd8c629be8e (CDE-888-Jira-Test)
Commit message: "CDE-888"
*First time build. Skipping changelog.*
[Checks API] No suitable checks publisher found.
[Pipeline] }
<p><span class="error">&#91;Pipeline&#93;</span> // stage
<span class="error">&#91;Pipeline&#93;</span> withEnv
<span class="error">&#91;Pipeline&#93;</span> {
 &gt; git config core.sparsecheckout # timeout=10
 &gt; git checkout -f a40df9adb4445594d996cb282389acd8c629be8e # timeout=10
<span class="error">&#91;Pipeline&#93;</span> stage
<span class="error">&#91;Pipeline&#93;</span> { (Test Build)</p>

<p> </p>

<p>This is the fourth build on a pull request, which is bizzard this is showing as "First Time Build." This causes errors downstream to other plugins:</p>

<p><span class="error">&#91;Pipeline&#93;</span> jiraSendDeploymentInfo</p>

<p>jiraSendDeploymentInfo: SKIPPED_ISSUE_KEYS_NOT_FOUND_AND_SERVICE_IDS_ARE_EMPTY: No issue keys found in the change log and service ids were not provided. Not sending deployment information to Jira: godeepseltzer.atlassian.net.</p>

<p> </p>

<p> </p>
</pre>
</details>

</ul>
</details>

<p>Hello, I am experiencing these issues as well using Bitbucket with a Multibranch Pipeline. This is causing issues with using other plugins looking to use the changelog, specifically the Atlassian Jira Cloud plugin.  Here is some logs on the issue:</p>

<p><span class="error">&#91;Pipeline&#93;</span> {<br/>
<span class="error">&#91;Pipeline&#93;</span> stage<br/>
<span class="error">&#91;Pipeline&#93;</span> </p>
{ (Declarative: Checkout SCM)
[Pipeline] checkout
The recommended git tool is: NONE
using credential Jenkins-Bitbucket-Build-Status-Notifier
Fetching changes from the remote Git repository
Fetching without tags
 > git rev-parse --is-inside-work-tree # timeout=10
 > git config remote.origin.url https://x-token-auth##@​bitbucket.org/##/jenkins-experimental.git # timeout=10
Fetching upstream changes from https://x-token-auth@​bitbucket.org/##/jenkins-experimental.git
 > git --version # timeout=10
 > git --version # 'git version 2.20.1'
 > git fetch --no-tags --force --progress -- https://x-token-auth:##@​bitbucket.org/##/jenkins-experimental.git +refs/heads/CDE-888-Jira-Test:refs/remotes/origin/CDE-888-Jira-Test # timeout=10
Checking out Revision a40df9adb4445594d996cb282389acd8c629be8e (CDE-888-Jira-Test)
Commit message: "CDE-888"
*First time build. Skipping changelog.*
[Checks API] No suitable checks publisher found.
[Pipeline] }
<p><span class="error">&#91;Pipeline&#93;</span> // stage<br/>
<span class="error">&#91;Pipeline&#93;</span> withEnv<br/>
<span class="error">&#91;Pipeline&#93;</span> {<br/>
 > git config core.sparsecheckout # timeout=10<br/>
 > git checkout -f a40df9adb4445594d996cb282389acd8c629be8e # timeout=10<br/>
<span class="error">&#91;Pipeline&#93;</span> stage<br/>
<span class="error">&#91;Pipeline&#93;</span> { (Test Build)</p>

<p> </p>

<p>This is the fourth build on a pull request, which is bizzard this is showing as "First Time Build." This causes errors downstream to other plugins:</p>

<p><span class="error">&#91;Pipeline&#93;</span> jiraSendDeploymentInfo</p>

<p>jiraSendDeploymentInfo: SKIPPED_ISSUE_KEYS_NOT_FOUND_AND_SERVICE_IDS_ARE_EMPTY: No issue keys found in the change log and service ids were not provided. Not sending deployment information to Jira: godeepseltzer.atlassian.net.</p>

<p> </p>

<p> </p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/markewaite.png" title="markewaite's avatar" /> markewaite:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-26354?focusedId=399606&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-399606">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=mkroell" class="user-hover" rel="mkroell">mkroell</a> you're describing a very different case than is described by the other comments in this issue.  This issue describes that the user would like the git plugin to provide a reliable guess of the commits that compose the first commit in a branch.  That causes the first build on a branch to show no changes.</p>

<p>I think the problem you're describing is that a non-initial build has changes but is not correctly calculating those changes.</p>
</pre>
</details>

</ul>
</details>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=mkroell" class="user-hover" rel="mkroell">mkroell</a> you're describing a very different case than is described by the other comments in this issue.  This issue describes that the user would like the git plugin to provide a reliable guess of the commits that compose the first commit in a branch.  That causes the first build on a branch to show no changes.</p>

<p>I think the problem you're describing is that a non-initial build has changes but is not correctly calculating those changes.</p>
